### PR TITLE
Proper Int64

### LINKED ISSUE #43: Field constructor vs static methods
Do we want stuff like `Field.fromBitstring()`?

### LINKED ISSUE #66: Finish `Int64` implementation

### PR DESCRIPTION
**Pull request by mitschabaude**
* Replace unfinished, unconstrained Int64 with a proper implementation
* Closes #66 
* Make Int64 API less boilerplate-y by allowing `number`, `string`, etc as method arguments
* Remove some hacks in toJson conversion, possible bc the new Int64 has the same layout as `Currency.Amount.Signed.t`

This is stacked on top of https://github.com/o1-labs/snarkyjs/pull/164, so that it can:
* Make previously failing Int64 unit tests work
* Add more Int64 unit tests

Since this will use updated bindings from https://github.com/MinaProtocol/mina/pull/11042, which fix an aspect of the `Field` constructor, this also adds a few new tests for the Field constructor.
* Closes #43

### CODE REVIEW

File: src/lib/int.ts
Comment (mitschabaude) [inline src/lib/int.ts:467]: TODO: add range checks (not constraints, just checks) on constant UInt64 / UInt32 / Int64 as you create them

Comment (mitschabaude) [inline src/lib/int.ts:467]: TODO: add range checks (not constraints, just checks) on constant UInt64 / UInt32 / Int64 as you create them

File: src/lib/int.ts
Comment (mimoo) [inline src/lib/int.ts:463]: why is this not implemented on `Int64`, something like `Int64.from(x: number | string | bigint | UInt64 | UInt32 | Int64)`

and same issue here, the `fromNumber` and other similar functions don't error if the user tries to input values that are too large (or too small)

File: src/lib/int.ts
Comment (mimoo) [inline src/lib/int.ts:45]: this return type is a bit error prone, why not return something like `{ quotient, rest }` instead?

Especially when we do things like `let  [q] =` or `let [, r] = `

File: src/lib/int.ts
Comment (mimoo) [inline src/lib/int.ts:444]: would `this.magnitude.equals(y_.magnitude).and(this.sign.equals(y_.sign)` be better? The two equalities could be optimized by the compiler (using the permutation), and thus you'd only have one constraint (the `and`)

**Review by mimoo — APPROVED** [review]
This is awesome! Code is clear, and this looks good to me, left a few minor comments

File: src/lib/int.ts
Comment (mitschabaude) [reply src/lib/int.ts:335]: good point, I'll add `private`, that makes it not visible to typescript (even though it's on the class)

File: src/lib/int.ts
Comment (mitschabaude) [reply src/lib/int.ts:358]: no, it also has `.neg()` at the end

File: src/lib/int.ts
Comment (mitschabaude) [reply src/lib/int.ts:380]: https://github.com/MinaProtocol/mina/blob/39af86e4a9f802277cdf73a6842c43fc4da48546/src/lib/snarky_js_bindings/lib/snarky_js_bindings_lib.ml#L1365

File: src/lib/int.ts
Comment (mitschabaude) [reply src/lib/int.ts:395]: I like it! I'll make `Field.one` the default argument for the sign

File: src/lib/int.ts
Comment (mitschabaude) [reply src/lib/int.ts:418]: It actually doesn't, when it goes through `fromFieldUnchecked` the range is checked.. I'll make sure that's done in all cases and add comments about it

File: src/lib/int.ts
Comment (mitschabaude) [reply src/lib/int.ts:463]: great suggestion!

File: src/lib/int.ts
Comment (mitschabaude) [reply src/lib/int.ts:444]: I didn't do that because I was worried about special-casing for 0, where -0 should equal 0.. The special-casing would also require multiplication constraints I think

File: src/lib/int.ts
Comment (mitschabaude) [reply src/lib/int.ts:419]: Yeah I think you may have a point.. I might reconsider this when range checks become cheap.

Do you mean shifting with `2^64 + res` and range-checking that? I don't follow how it works with multiplication & negative numbers.

Currently, it's convenient to have Int64 be {magnitude, sign} because that mirrors the Ocaml data structure. If I make it a single Field constrained to 64 bits, I allow a slightly smaller range than the Ocaml type (64 vs 65 bits)

File: src/lib/int.ts
Comment (mimoo) [reply src/lib/int.ts:358]: 🤦🏻‍♂️

File: src/lib/int.ts
Comment (mimoo) [reply src/lib/int.ts:419]: yeah `+` not `*` sorry

File: src/lib/int.ts
Comment (mimoo) [reply src/lib/int.ts:444]: good point :o

File: src/lib/int.ts
Comment (mimoo) [reply src/lib/int.ts:380]: 🙏🏻

**Comment by mitschabaude** [comment]
@mimoo addressed all feedback,
also added Field.to/fromBigint() to make the code cleaner in some places

this should be good to go, will merge once the parent PR is merged

**Pull request merged**
